monix-reactive/shared/src/main/scala/monix/reactive/internal/operators/IntersperseObservable.scala

```diff
@@ -0,0 +1,56 @@
+package monix.reactive.internal.operators
+
+import monix.execution.Ack.Continue
+import monix.execution.atomic.AtomicBoolean
+import monix.execution.{Ack, Cancelable}
+import monix.reactive.Observable
+import monix.reactive.observers.Subscriber
+
+import scala.concurrent.Future
+
+/**
+  * Created by omainegra on 5/29/17.
+  */
+private[reactive] final class IntersperseObservable[+A](source: Observable[A],
+                                                        separator: A) extends Observable[A]{ self =>
+
+  override def unsafeSubscribeFn(out: Subscriber[A]): Cancelable = {
+    val firstTime = AtomicBoolean(true)
+    val isDone = AtomicBoolean(false)
+    var downstreamAck = Continue : Future[Ack]
+
+    val upstream = source.unsafeSubscribeFn(new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      override def onNext(elem: A): Future[Ack] = {
+        downstreamAck = if (firstTime.getAndSet(false)) {
+          out.onNext(elem)
+        }
+        else {
+          out.onNext(separator).flatMap {
```

Comment:
Doing a `flatMap` on a `Future` can be pretty inefficient since it forces an asynchronous boundary.

This is why Monix introduces a couple of special Future extension methods for avoiding async boundaries if the `Future` returned by `onNext` is already complete. Monix is efficient because it tries to collapse these async boundaries if the downstream subscriber returns a plain `Continue` or `Stop` that's immediately available.

Don't get me wrong, this code is correct, but it can be more efficient if you use `syncFlatMap` instead:

```scala
out.onNext(separator).syncFlatMap {
  ???
}
```

As a note, for user code this is actually dangerous because it is stack-unsafe (in loops), but we know what we are doing here and it's fine.